paasta_tools/cli/cmds/rollback.py

```diff
@@ -65,57 +60,53 @@ def add_subparser(subparsers):
     list_parser.set_defaults(command=paasta_rollback)
 
 
-def validate_given_instances(service_instances, args_instances):
-    """Given two lists of instances, return the intersection and difference between them.
+def validate_given_deploy_groups(service_deploy_groups, args_deploy_groups):
+    """Given two lists of deploy groups, return the intersection and difference between them.
 
-    :param service_instances: instances actually belonging to a service
-    :param args_instances: the desired instances
-    :returns: a tuple with (common, difference) indicating instances common in both
-    lists and those only in args_instances
+    :param service_deploy_groups: instances actually belonging to a service
+    :param args_deploy_groups: the desired instances
+    :returns: a tuple with (common, difference) indicating deploy groups common in both
+    lists and those only in args_deploy_groups
     """
-    if len(args_instances) is 0:
-        valid_instances = set(service_instances)
-        invalid_instances = set([])
+    if len(args_deploy_groups) is 0:
+        valid_deploy_groups = set(service_deploy_groups)
+        invalid_deploy_groups = set([])
     else:
-        valid_instances = set(args_instances).intersection(service_instances)
-        invalid_instances = set(args_instances).difference(service_instances)
+        valid_deploy_groups = set(args_deploy_groups).intersection(service_deploy_groups)
+        invalid_deploy_groups = set(args_deploy_groups).difference(service_deploy_groups)
 
-    return valid_instances, invalid_instances
+    return valid_deploy_groups, invalid_deploy_groups
 
 
 def paasta_rollback(args):
     """Call mark_for_deployment with rollback parameters
     :param args: contains all the arguments passed onto the script: service,
-    cluster, instance and sha. These arguments will be verified and passed onto
+    deploy groups and sha. These arguments will be verified and passed onto
     mark_for_deployment.
     """
     service = figure_out_service_name(args)
-    cluster = args.cluster
     git_url = get_git_url(service)
     commit = args.commit
-    given_instances = args.instances.split(",")
+    given_deploy_groups = filter(bool, args.deploy_groups.split(","))
```

Comment:
The `instances` parameter to paasta rollback is supposed to be optional; if it's not included rollback should rollback every instance.

This code removes empty strings from `given_instances`. Before this if the instances argument wasn't passed to rollback it would default to `''`, so given_instances would be `['']`. This causes rollback not to work properly as the rest of the code checks to see if `len(given_instances) == 0`.

I chose to make this change rather than changing the logic for determining if `given_instances` is empty because this also deals with some typos (correcting multiple commas like `'instance1,,instance2'`) and also removing trailing commas (before `'instance_name,'` would resolve to `['instance_name', '']`).

TL;DR this feature was always broken and filtering with the `bool` function removes empty strings from the result of `split`.